drivers/pwm/pwm_rcar.c

```diff
@@ -0,0 +1,279 @@
+#define RCAR_PWM_DIVISER_MASK       (RCAR_PWM_CR_CC_MASK | RCAR_PWM_CR_CCMD)
+#define RCAR_PWM_DIVISER_SHIFT      15
+
+/* PWMCNT (PWM Count Register) */
+#define RCAR_PWM_CNT_CYC_MASK       0x03ff0000    /* PWM Cycle */
+#define RCAR_PWM_CNT_CYC_SHIFT      16
+#define RCAR_PWM_CNT_PH_MASK        0x000003ff    /* PWM High-Level Period */
+#define RCAR_PWM_CNT_PH_SHIFT       0
+
+struct pwm_rcar_cfg {
+	uint32_t reg_addr;
+	const struct device *clock_dev;
+	struct rcar_cpg_clk core_clk;
+	struct rcar_cpg_clk mod_clk;
+	const struct pinctrl_dev_config *pcfg;
+};
+
+struct pwm_rcar_data {
+	uint32_t clk_rate;
+};
+
+static uint32_t pwm_rcar_read(const struct pwm_rcar_cfg *config,
+			      uint32_t offs)
+{
+	return sys_read32(config->reg_addr + offs);
+}
+
+static void pwm_rcar_write(const struct pwm_rcar_cfg *config,
+			   uint32_t offs, uint32_t value)
+{
+	sys_write32(value, config->reg_addr + offs);
+}
+
+static void pwm_rcar_write_bit(const struct pwm_rcar_cfg *config,
+			       uint32_t offs, uint32_t bits, bool value)
+{
+	uint32_t reg_val = pwm_rcar_read(config, offs);
+
+	if (value) {
+		reg_val |= bits;
+	} else {
+		reg_val &= ~(bits);
+	}
+
+	pwm_rcar_write(config, offs, reg_val);
+}
+
+static int pwm_rcar_update_clk(const struct pwm_rcar_cfg *config, uint32_t pwm,
+			       uint32_t *period_cycles, uint32_t *pulse_cycles)
+{
+	uint32_t reg_val, power, diviser;
+
+	power = pwm_rcar_read(config, RCAR_PWM_CR) & RCAR_PWM_DIVISER_MASK;
+	power = power >> RCAR_PWM_DIVISER_SHIFT;
+	diviser = 1 << power;
+
+	LOG_DBG("Found old diviser : 2^%d=%d", power, diviser);
+
+	/* Looking for the best possible clock diviser */
+	if (*period_cycles > RCAR_PWM_MAX_CYCLE) {
+		/* Reducing clock speed */
+		while (*period_cycles > RCAR_PWM_MAX_CYCLE) {
+			diviser *= 2;
+			*period_cycles /= 2;
+			*pulse_cycles /= 2;
+			power++;
+			if (power > RCAR_PWM_MAX_DIV) {
+				return -ENOTSUP;
+			}
+		}
+	} else {
+		/* Increasing clock speed */
+		while (*period_cycles < (RCAR_PWM_MAX_CYCLE / 2)) {
+			if (power == 0) {
+				return -ENOTSUP;
+			}
+			diviser /= 2;
+			*period_cycles *= 2;
+			*pulse_cycles *= 2;
+			power--;
+		}
+	}
+	LOG_DBG("Found new diviser : 2^%d=%d\n", power, diviser);
+
+	/* Set new clock Diviser */
+	reg_val = pwm_rcar_read(config, RCAR_PWM_CR);
+	reg_val &= ~RCAR_PWM_DIVISER_MASK;
+	reg_val |= (power << RCAR_PWM_DIVISER_SHIFT);
+	pwm_rcar_write(config, RCAR_PWM_CR, reg_val);
+
+	return 0;
+}
+
+static int pwm_rcar_pin_set(const struct device *dev, uint32_t pwm,
+			    uint32_t period_cycles, uint32_t pulse_cycles,
+			    pwm_flags_t flags)
+{
+	const struct pwm_rcar_cfg *config = dev->config;
+	uint32_t reg_val;
+	int ret = 0;
+
+	/* Assume only one channel per pwm instance */
+	if (pwm != 0) {
+		return -ENOTSUP;
+	}
+
+	if (flags != PWM_POLARITY_NORMAL) {
+		return -ENOTSUP;
+	}
+
+	/* Prohibiten values */
+	if (period_cycles == 0U || pulse_cycles == 0U ||
+	    pulse_cycles > period_cycles) {
+		return -EINVAL;
```

Comment:
ENOTSUP as it is not supported by the pwn controller ?

Reply:
0 is a prohibited value so "Unsupported value" is preferred than "Invalid argument". Correct. Fixed. Thanks :)